### DISCUSSION TITLE
How to sequentially call fit() and test() in DDP

### DISCUSSION THREAD

**Discussion by kapsner** [Lightning-AI/pytorch-lightning#6354] [DDP / multi-GPU / multi-node] [answered]
I wrote a Python script that loops over sets of hyperparameters and for each set calls `trainer.fit()`. Subsequently `trainer.test()` is currently called 2 times, each for the best checkpoint that was logged for two metrics.

This script is executable via commandline and works well on CPU an 1 GPU for any number of hyperparameter sets.

I want to run this code on multiple GPUs (2-4) and wonder if I could use 'ddp'?

The [latest documentation](https://pytorch-lightning.readthedocs.io/en/latest/advanced/multi_gpu.html) says that if `your script needs to invoke both .fit and .test, or one of them multiple times` 'ddp' mode isn't possible.

However, @awaelchli indicated [here](https://forums.pytorchlightning.ai/t/how-to-run-trainer-fit-and-trainer-test-in-ddp-distributed-mode/192/6), that running `trainer.fit()` and `trainer.test()` in 'ddp' mode is now possible but the documentation is outdated.

Questions:
* Is it now possible to run both  `trainer.fit()` and subsequently `trainer.test()` (and eventually `trainer.test()` mutiple times) in 'ddp' mode?
* Would I need to reorganize my code so that the loop over hyperparameter sets invokes a `subprocess.call` to actually instantiate every new trainer for 'ddp' in another process?
* If calling `.test()` after `.fit()` in 'ddp' is still not possible, would it be a workaround to instantiate `trainer` again after `.fit()` is finished and just load the model weights from the best checkpoint for testing (to avoid using the same trainer object)?

I would be really happy for some help on that.

**Comment by awaelchli** [comment, marked as answer]
No this should be working fine. 
We have tests for this here: https://github.com/PyTorchLightning/pytorch-lightning/blob/a6c98c4e4944cfc4977d27a4ea1e0c7c621dfb4d/tests/accelerators/test_ddp.py#L56
I will update the docs.

**Reply by kapsner** [reply]
Ok, thx - I will figure out how to get this running for me!
